### PR TITLE
feat(react): add castle 2.0 analytics' integration

### PR DESCRIPTION
**Pull request by ruifcnunes**
## Description
This PR adds a new analytics integration: Castle v2.x.
In order to avoid breaking changes, the current Castle V1 class was used as a facade to load the correct castle version depending on the options passed: if passed the `appId`, it loads the current V1 integration. If passed the new `pk`(publishable key), it loads the new V2 version.

## Checklist

- [x] The commit message follows our guidelines
- [x] Tests for the respective changes have been added
- [x] The code is commented, particularly in hard-to-understand areas
- [x] The labels and/or milestones were added

## Disclaimer

By sending us your contributions, you are agreeing that your contribution is made subject to the terms of our [Contributor Ownership Statement](https://github.com/Farfetch/.github/blob/master/COS.md)

### CODE REVIEW

File: packages/react/src/analytics/integrations/Castle/CastleV2.js [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+  /**
+   * Callback that is used on the Axios interceptor to handle any error when setting the Castle header.
+   *
+   * @async
+   *
+   * @param {object|string} error - The error itself.
+   *
+   * @returns {Promise} - The rejected promise.
+   */
+  onBeforeRequestReject = async error => {
```
Comment (boliveira) [inline packages/react/src/analytics/integrations/Castle/CastleV2.js:187]: This method can be removed as by default axios will reject the promise automatically if there was an error with the request.
To do so, you can change the attach interceptors instruction to this:

```js
// install the default interceptors on our core client
    this.httpClientInterceptor = this.httpClient.interceptors.request.use(
      this.onBeforeRequestFullfil,
      null (or undefined),
    );

```

File: packages/react/src/analytics/integrations/Castle/CastleV2.js [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+        };
+
+        return this.castleJS.form(formData).then(
+          () =>
+            this.debugModeOn &&
+            utils.logger.info(
+              `${CASTLE_MESSAGE_PREFIX} Form track success. Payload:`,
+              formData,
+            ),
+          error => Promise.reject(error),
```
Comment (boliveira) [inline packages/react/src/analytics/integrations/Castle/CastleV2.js:270]: This is the same case, you can remove this function argument and the promise will be rejected automatically.

```js
return this.castleJS.form(formData).then(
          () =>
            this.debugModeOn &&
            utils.logger.info(
              `${CASTLE_MESSAGE_PREFIX} Form track success. Payload:`,
              formData,
            ),
         );
    }
```

File: packages/react/src/analytics/integrations/Castle/CastleV2.js [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+        };
+
+        return this.castleJS.custom(eventData).then(
+          () =>
+            this.debugModeOn &&
+            utils.logger.info(
+              `${CASTLE_MESSAGE_PREFIX} Custom event track success. Payload:`,
+              eventData,
+            ),
+          error => Promise.reject(error),
```
Comment (boliveira) [inline packages/react/src/analytics/integrations/Castle/CastleV2.js:288]: Same as https://github.com/Farfetch/blackout/pull/254/files#r842570201

File: packages/react/src/analytics/integrations/Castle/CastleV2.js [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+        return;
+    }
+  }
+
+  /**
+   * Method that will handle each type of event to be tracked.
+   *
+   * @param {object} data - Data Track event data.
+   *
+   * @returns {Promise<Castle>} - The resolved promise of each castle call method.
```
Comment (boliveira) [inline packages/react/src/analytics/integrations/Castle/CastleV2.js:242]: Does the call to `this.castleJS.form` return the `Castle` instance?

File: packages/react/src/analytics/integrations/Castle/CastleV2.js [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+    };
+
+    return this.castleJS.page(pageData).then(
+      () =>
+        this.debugModeOn &&
+        utils.logger.info(
+          `${CASTLE_MESSAGE_PREFIX} Page track success. Payload:`,
+          pageData,
+        ),
+      error => Promise.reject(error),
```
Comment (boliveira) [inline packages/react/src/analytics/integrations/Castle/CastleV2.js:323]: Same as https://github.com/Farfetch/blackout/pull/254/files#r842571224

File: packages/react/src/analytics/integrations/Castle/CastleV2.js [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+
+  /**
+   * Method that will handle page events to be tracked.
+   *
+   * @param {object} data - Data Track event data.
+   *
+   *
+   * @async
+   *
+   * @returns {Promise<Castle>} - The resolved promise of each castle call method.
```
Comment (boliveira) [inline packages/react/src/analytics/integrations/Castle/CastleV2.js:302]: Same as https://github.com/Farfetch/blackout/pull/254/files#r842572824

**Review by boliveira — APPROVED** [review]

**Review by danielbento92 — APPROVED** [review]

**Pull request merged**